ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,862 @@
+    public_keys: &[SigmaBoolean],
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let tx = tx_context.spending_tx.clone();
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = tx_context
+            .get_boxes_to_spend()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let ctx = Rc::new(make_context(state_context, &tx_context, i)?);
+        let tree = input_box.ergo_tree.clone();
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(i, generate_commitments_for(sigma_tree, public_keys));
+    }
+    Ok(hints_bag)
+}
+
+/// Extracting hints from a transaction and outputs it's corresponding TransactionHintsBag
+pub fn extract_hints(
+    signed_tx: Transaction,
+    tx_context: TransactionContext,
+    state_context: &ErgoStateContext,
+    real_secrets_to_extract: Vec<SigmaBoolean>,
+    simulated_secrets_to_extract: Vec<SigmaBoolean>,
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let mut hints_bag = TransactionHintsBag::empty();
+    let tx = tx_context.spending_tx.clone();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = tx_context
+            .get_boxes_to_spend()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let ctx = Rc::new(make_context(state_context, &tx_context, i)?);
+        let tree = input_box.ergo_tree.clone();
+        let proof: ProofBytes = signed_tx
+            .inputs
+            .get(i)
+            .unwrap()
+            .clone()
+            .spending_proof
+            .proof;
+        let proof: Vec<u8> = Vec::from(proof);
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(
+            i,
+            bag_for_multi_sig(
+                sigma_tree,
+                real_secrets_to_extract.as_slice(),
+                simulated_secrets_to_extract.as_slice(),
+                proof.as_slice(),
+            ),
+        );
+    }
+
+    Ok(hints_bag)
+}
+
+#[allow(clippy::unwrap_used)]
+/// Extracting hints from reduced transaction
+pub fn extract_hints_from_reduced_transaction(
```

Comment:
Why do we need hint extraction for reduced tx? It's an unsigned tx while hint extraction requires a signed tx.